ethpm/manifest.py

```diff
@@ -0,0 +1,149 @@
+    Class for handling manifest generation that represents a set of packaged contract(s).
+
+    Expected directory structure:
+        |-- combined.json (solc output)
+        |-- ... (will write manifests to disk here) ...
+        |-- contracts/
+        |   |-- Contract1.sol
+        |   |-- Contract2.sol
+    """
+
+    def __init__(self, manifest_version: str, package_name: str, version: str) -> None:
+        validate_manifest_version(manifest_version)
+        validate_package_name(package_name)
+        self.manifest_version = manifest_version
+        self.package_name = package_name
+        self.version = version
+        self.ipfs_backend = get_ipfs_backend()
+        self.solc_path = None  # type: Path
+        self.sources = {}  # type: Dict[str,str]
+        self.contract_types = {}  # type: Dict[str,Any]
+
+    def link_solc_output(
+        self, solc_path: Path, contract_types: List[str] = None
+    ) -> None:
+        """
+        Populate the manifest with data from a linked solc output.
+
+        :solc_path
+            Path leading to a file containing output from solc
+            Solc output *must* be generated with `abi` & `devdoc` flags enabled
+
+            i.e.
+        `solc --output-dir ./ --combined-json abi,bin,devdoc,interface,userdoc contracts/Owned.sol`
+
+        :contract_types
+            List containing the contract types to be included in this manifest.
+
+            * Packages should only include contract types that can be
+              found in the source files for this package.
+            * Packages should not include abstract contracts in the contract types section.
+            * Packages should not include contract types from dependencies.
+        """
+        if self.solc_path:
+            raise AttributeError(
+                "This manifest has already been linked to solc output located at {0}.".format(
+                    self.solc_path
+                )
+            )
+        self.solc_path = solc_path
+        solc = load_json_from_file_path(self.solc_path)
+        for contract in solc["contracts"]:
+            contract_path, contract_name = contract.split(":")
+            source_data = self._create_source(contract_path)
+            setattr(self, "sources", {**self.sources, **source_data})
+            if contract_types is not None and contract_name in contract_types:
+                contract_type_data = self._create_contract_type(
+                    contract_name, contract, solc
+                )
+                setattr(
+                    self,
+                    "contract_types",
+                    {**self.contract_types, **contract_type_data},
+                )
+
+    @to_dict
+    def _create_contract_type(
+        self, contract_name: str, contract: str, solc: Dict[str, Any]
+    ) -> Generator[Tuple[str, Any], None, None]:
+        yield contract_name, self._generate_contract_type_dict(contract, solc)
+
+    @to_dict
+    def _generate_contract_type_dict(
```

Comment:
All of these that don't actually use `self` are ones I'd recommend moving to utility functions (just here in this module is fine).